### PR TITLE
Fix/64 - Add admin webclient

### LINKED ISSUE #65: CronChannel doesn't use loop argument at startup
When executing test with CronChannel, main loop is removed from thread then aicron doesn't receive the loop and try to get the main one and... fails.

### PR DESCRIPTION
**Pull request by jrmi**
- But authentication is missing for now so don't use in production yet.
- Also fix #65 
- Message store function are now async

### CODE REVIEW

File: pypeman/msgstore.py [hunk 1]
```diff
@@ -177,47 +219,81 @@ def store(self, msg):
...
+        position = 0
+        for year in await self.sorted_list_directories(os.path.join(self.base_path), reverse=reverse):
+            for month in await self.sorted_list_directories(os.path.join(self.base_path, year), reverse=reverse):
+                for day in await self.sorted_list_directories(os.path.join(self.base_path, year, month), reverse=reverse):
                     for msg_name in sorted(os.listdir(os.path.join(self.base_path, year, month, day)), reverse=reverse):
                         found = self.msg_re.match(msg_name)
                         if found:
-                            id = os.path.join(year, month, day, msg_name)
-                            yield self.get(id)
+                            if start <= position and position < end:
```
Comment (feenes) [inline pypeman/msgstore.py]: `if start <= position < end:`

File: pypeman/remoteadmin.py [hunk 2]
```diff
@@ -218,16 +242,21 @@ def stop(self, channel):
         """
         return self.send_command('stop_channel', [channel])
 
-    def list_msg(self, channel):
+    def list_msg(self, channel, start=0, count=10, order_by='timestamp'):
         """
         List first 10 messages on specified channel from remote instance.
```
Comment (feenes) [inline pypeman/remoteadmin.py]: list 'count' first messages on . . .

File: pypeman/remoteadmin.py [hunk 3]
```diff
@@ -267,6 +306,7 @@ def wrapper(self, *arg, **kwargs):
     return wrapper
 
 
+
```
Comment (feenes) [inline pypeman/remoteadmin.py]: lines

File: pypeman/tests/test_msgstore.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+        print("Process %s" % self.name)
+
+        if msg.timestamp.day == 12:
+            raise TestException()
+
+        return msg
+
+
+class ExceptNode(TestNode):
+    # This node raise an exception
```
Comment (feenes) [inline pypeman/tests/test_msgstore.py]: raises

File: pypeman/msgstore.py [hunk 5]
```diff
@@ -112,28 +129,40 @@ def __init__(self, base_dict, store_id):
...
         resp = dict(self.messages[id])
         resp['message'] = Message.from_dict(resp['message'])
         return resp
 
-    def search(self, order_by='timestamp'):
+    async def search(self, start=0, count=10, order_by='timestamp'):
+
+        if order_by.startswith('-'):
+            reverse = True
+            sort_key = order_by[1:]
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/msgstore.py]: I copy and prefer django notation ?

File: pypeman/remoteadmin.py [hunk 6]
```diff
@@ -37,9 +41,11 @@ class RemoteAdminServer():
     Expose json/rpc function to a client by a websocket.
     """
 
-    def __init__(self, host, port):
+    def __init__(self, loop=None, host='localhost', port='8091', ssl=None):
         self.host = host
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/remoteadmin.py]: Settings are used in call !

File: pypeman/remoteadmin.py [hunk 7]
```diff
@@ -164,9 +192,14 @@ def get_channel(self, name):
...
 
 class RemoteAdminClient():
-    def __init__(self, url):
+    """
+    Remote admin client. To be use by ipython shell or pypeman shell.
+
+    :params url: Pypeman Websocket url.
+    """
+    def __init__(self, loop=None, url='ws://localhost:8091'):
         self.url = url
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/remoteadmin.py]: Same

File: pypeman/channels.py [hunk 8]
```diff
@@ -108,6 +109,8 @@ def __init__(self, name=None, parent_channel=None, loop=None, message_store_fact
 
         self.message_store = self.message_store_factory.get_store(self.name)
 
+        self._first_start = True
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/channels.py]: It's really first start of channel.

File: pypeman/channels.py [hunk 9]
```diff
@@ -64,6 +64,7 @@ def __init__(self, name=None, parent_channel=None, loop=None, message_store_fact
         self._nodes = []
         self._node_map = {}
         self._status = BaseChannel.STOPPED
+        self.processed = 0
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/channels.py]: Ok

File: pypeman/channels.py [hunk 10]
```diff
@@ -356,11 +363,54 @@ def handle_and_wait(self, msg):
...
+            'status': BaseChannel.status_id_to_str(self.status),
+            'has_message_store': not isinstance(self.message_store, msgstore.NullMessageStore),
+            'processed': self.processed,
+        }
+
+    def subchannels(self):
+        res = []
+
+        for node in self._nodes:
+            if isinstance(node, SubChannel):
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/channels.py]: Totally.

File: pypeman/channels.py [hunk 11]
```diff
@@ -356,11 +363,54 @@ def handle_and_wait(self, msg):
...
+                chan_dict['subchannels'] = node.subchannels()
+                res.append(chan_dict)
+            elif isinstance(node, Case):
+                for cond, channel in node.cases:
+                    chan_dict = channel.to_dict()
+                    chan_dict['subchannels'] = channel.subchannels()
+                    res.append(chan_dict)
+        return res
+
+    def flat_subchannels(self):
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/channels.py]: In fact flat_subchannels was the first version unused anymore. I prefer to remove this code, what do you think. If someone want a flat version he can flatten it by himself.

File: pypeman/client/.babelrc [hunk 12]
```diff
@@ -0,0 +1,18 @@
+{
+  "presets": [
+    ["env", {
+      "modules": false,
+      "targets": {
+        "browsers": ["> 1%", "last 2 versions", "not ie <= 9"]
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/client/.babelrc]: No, compatibility with browser that market share is > 1%

File: pypeman/client/.editorconfig [hunk 13]
```diff
@@ -0,0 +1,9 @@
+root = true
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/client/.editorconfig]: Already discussed.

File: pypeman/remoteadmin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Zluurk) [reply pypeman/remoteadmin.py]: why have default settings ?

File: pypeman/client/src/components/ChannelList.vue [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+    },
+    showError (err) {
+      console.log('Error while calling server', err)
+    }
+  },
+  data () {
+    return {
+      headers: [
+        {text: 'Channel name', align: 'left', value: 'name'},
+        {text: 'Msg Processed', value: 'processed'},
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/client/src/components/ChannelList.vue]: Ok

File: pypeman/msgstore.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zluurk) [reply pypeman/msgstore.py]: ok

File: pypeman/commands.py [hunk 15]
```diff
@@ -16,9 +16,10 @@
 if len(sys.argv) == 1:
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/commands.py]: Still :cry:

File: pypeman/msgstore.py [hunk 16]
```diff
@@ -85,16 +99,19 @@ def get_store(self, store_id):
...
-    def search(self, order_by='timestamp'):
+    async def search(self, **kwargs):
         return []
 
+    async def total(self):
+        return 0
+
 
 class MemoryMessageStoreFactory(MessageStoreFactory):
     """ Return a Memory message store. All message are loose at pypeman stop. """
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/msgstore.py]: Did I wrote this ???

File: pypeman/msgstore.py [hunk 17]
```diff
@@ -177,47 +219,81 @@ def store(self, msg):
...
         """
         return sorted([d for d in os.listdir(path) if os.path.isdir(os.path.join(path, d))], reverse=reverse)
 
-    def search(self, order_by='timestamp'):
-        reverse = (order_by == 'timestamp')
-
-        for year in self.sorted_list_directories(os.path.join(self.base_path), reverse=reverse):
-            for month in self.sorted_list_directories(os.path.join(self.base_path, year), reverse=reverse):
-                for day in self.sorted_list_directories(os.path.join(self.base_path, year, month), reverse=reverse):
+    async def count_msg(self):
```
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/msgstore.py]: Ok

File: pypeman/msgstore.py (hunk 1 above)
Comment (jrmi) [reply pypeman/msgstore.py]: Not sure it was working. I love that.

File: pypeman/remoteadmin.py (hunk 2 above)
Comment (jrmi) [reply pypeman/remoteadmin.py]: Ok

File: pypeman/tests/test_msgstore.py (hunk 4 above)
Comment (jrmi) [reply pypeman/tests/test_msgstore.py]: Again !

File: pypeman/remoteadmin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jrmi) [reply pypeman/remoteadmin.py]: I use them in tests for example.

File: pypeman/remoteadmin.py (hunk 2 above)
Comment (jrmi) [reply pypeman/remoteadmin.py]: done : 4a4fe320362e44320ef8eef8f5df0448c3923f33

File: pypeman/remoteadmin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Zluurk) [reply pypeman/remoteadmin.py]: ok

File: pypeman/channels.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (feenes) [reply pypeman/channels.py]: agreed. create it when really needed and code it then such, that it is DRY

**Pull request merged**